src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -69,12 +87,63 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
 	/* ensure that the task is not NULL */
 	Assert(task != NULL);
 
-	/* disallow transactions and triggers during distributed modify commands */
+	/* disallow triggers during distributed modify commands */
 	if (commandType != CMD_SELECT)
 	{
-		bool topLevel = true;
-		PreventTransactionChain(topLevel, "distributed commands");
 		eflags |= EXEC_FLAG_SKIP_TRIGGERS;
+
+		if (IsTransactionBlock() && xactParticipantHash == NULL)
```

Comment:
> > Hm. What about checking `QueryDesc->dest`, and treating everything but `DestRemote`, `DestRemoteExecute` as commands needing a transaction? I think that should do what we want.
> 
> _This_ I do not understand. Can you clarify what you're proposing? It sounds clean and like it could handle the other "multi-statements-inside-transactions" cases (functions and multi-command strings), but I don't know what `QueryDesc->dest` has to do with those cases. Quick explanation?

->dest is the destination where the results of a query a directed to. As
it happens, '*Remote' are the ones used when sending the results of a
query directly to the client (via the socket). The rest are used when
the results instead are received by other methods - and those pretty
much happen only when being called via a function, inside COPY, and such
stuff.  If we use a transaction for everything but *Remote, we have the
'function' use-case nailed, without incuring the overhead for
transaction-less commands from the client directly.

Makes more sense now?